imports/server/hooks/ChatNotificationHooks.ts

```diff
@@ -0,0 +1,105 @@
+
+const ChatNotificationHooks: Hookset = {
+  async onChatMessageCreated(chatMessageId: string) {
+    // This method implements notifications for both Dingwords and @-mentions together,
+    // so that we do not generate duplicate notifications if a user is both @-mentioned
+    // and one of their dingwords is spoken in the same breath.
+
+    // Dingwords can be disabled (they're comparatively expensive, since they do big-O(users) work),
+    // but @-mentions cannot (they're cheap, doing big-O(message components) work).
+
+    const chatMessage = (await ChatMessages.findOneAsync(chatMessageId))!;
+
+    // Collect users to notify into a set, and then create notifications at the end.
+    const usersToNotify = new Set<string>();
+    if (!chatMessage.sender) {
+      // Don't notify for system messages.
+      return;
+    }
+
+    // Notify for @-mentions in message.
+    if (chatMessage.content) {
+      await Promise.all(chatMessage.content.children.map(async (child) => {
+        if (nodeIsMention(child)) {
+          const mentionedUserId = child.userId;
+          if (mentionedUserId !== chatMessage.sender) {
+            // Don't have messages notify yourself.
+            const mentionedUser = await MeteorUsers.findOneAsync(mentionedUserId);
+            if (mentionedUser?.hunts?.includes(chatMessage.hunt)) {
+              // Only allow mentions of users that are in the current hunt.
+              usersToNotify.add(mentionedUserId);
+            }
+          }
+        }
+      }));
+    }
+
+    // Respect feature flag.
+    if (!Flags.active('disable.dingwords')) {
+      const normalizedText = chatMessage.text?.trim().toLowerCase() ??
+        chatMessage.content?.children.map((child) => {
+          if (nodeIsText(child)) {
+            return child.text;
+          } else {
+            // No need to look for dingwords in @-mentions, but let them split words
+            return ' ';
+          }
+        }).join('').trim().toLowerCase() ?? '';
+
+      // Find all users who are in this hunt with dingwords set.
+      for await (const u of MeteorUsers.find({
+        hunts: chatMessage.hunt,
+        'dingwords.0': { $exists: true },
+      }, {
+        fields: { _id: 1, dingwords: 1 },
+      })) {
+        // Avoid making users ding themselves.
+        if (u._id === chatMessage.sender) {
+          // eslint-disable-next-line no-continue
+          continue;
+        }
+
+        const dingwords = u.dingwords;
+        if (dingwords) {
+          const matches = dingwords.some((dingword) => normalizedText.includes(dingword));
+          if (matches) {
+            await ChatNotifications.insertAsync({
+              user: u._id,
+              sender: chatMessage.sender,
+              puzzle: chatMessage.puzzle,
+              hunt: chatMessage.hunt,
+              text: chatMessage.text,
+              content: chatMessage.content,
+              timestamp: chatMessage.timestamp,
+            });
```

Comment:
Despite the comment above, this actually does generate two separate notifications in the case of a coincident mention and ding. Since it looks like the intention is for both types of the notification to be the same (that is, one style isn't supposed to overrule the other), I think you just meant:
```suggestion
            usersToNotify.add(u._id);
```

Reply:
Quite right.  I might have messed this up in one of the rebases I did?  or maybe it was just buggy from the get-go.  Either way, good catch, and thanks!  Fixed in 27fcf78486573dbb0bc215ada2594c63975e8fe3